xapian-applications/omega/handler_gmime.cc

```diff
@@ -0,0 +1,174 @@
+/** @file handler_gmime.cc
+ * @brief Extract text and metadata using gmime.
+ */
+/* Copyright (C) 2019 Bruno Baruffaldi
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "myhtmlparse.h"
+#include "utf8convert.h"
+#include <string.h>
+
+#include <glib.h>
+#include <gmime/gmime.h>
+
+using namespace std;
+
+const unsigned SIZE = 4096;
+
+static void
+extract_html(const string& text, string& charset, string& dump)
+{
+    MyHtmlParser parser;
+    if (charset.empty())
+	charset = "UTF-8";
+    try {
+	parser.ignore_metarobots();
+	parser.parse_html(text, charset, false);
+	dump += parser.dump;
+    } catch (const string& newcharset) {
+	parser.reset();
+	parser.ignore_metarobots();
+	parser.parse_html(text, newcharset, true);
+	dump += parser.dump;
+    } catch (...) {
+	return;
+    }
```

Comment:
Quietly ignoring all unexpected exceptions seems a bad idea.  If there particular exceptions that we should be ignoring, be more specific.  If not, I'd drop the default exception handling here.

Reply:
I am not sure if I should drop the default exception handling here. If there was any exception(different from newcharset) inside `extract_html`, it would return without appending anything to dump.
Maybe I can add some exception handling to `parser_content` at `extract`, and if there is any exception, the dump will be ignored (and return just the `author` and the `title`).

Reply:
Quietly ignoring an unexpected exception at any level is rarely helpful.  It might seem like you're writing more robust code by doing so, but what it actually means is that nobody's likely to report or debug a problem which results in an exception being thrown because they simply don't see any evidence of it (unless they look closely enough to notice some document content missing, but people rarely do unless they are looking for a particular affected document they know for sure is present but discover they can't find it in the search).

If you're going to handle unexpected exceptions at all just do it at the top level like the other helpers and return an error across the connection - e.g. for poppler you have:

```
    } catch (...) {
        error = "Poppler threw an exception";
        return false;
    } 
```

If there are particular exceptions that happen in the normal course of processing documents with a particular library, then they can be considered on a case-by-case basis - in some cases it might be reasonable to just treat them like an empty field.  But unless we know of such exceptions it's better to default to reporting exceptions as then we will at least find out about them.